File: ggcommons/metrics/targets/cloudwatch.py
import boto3
import time
from threading import Thread, Event
from ggcommons.config.manager.config_manager import ConfigManager
from ggcommons.metrics.targets.metric_target import MetricTarget


class CloudWatch(MetricTarget):
    # CloudWatch PutMetricData accepts at most 1000 metric datums per request.
    MAX_DATUMS_PER_REQUEST = 1000

    def __init__(self, config_manager: ConfigManager):
        super().__init__(config_manager)
        self.logger.info("Initializing CloudWatch metric target")
        
        self._cloudwatch_client = boto3.client("cloudwatch", region_name="us-east-1")
        self._pending_metrics = {}
        self._interval_secs = -1
        self._flush_event = None
        self._flush_thread = None
        self._terminate_thread = False
        
        self.logger.info(f"CloudWatch client initialized for region: us-east-1")
        self._start_periodic_flush()

    def _start_periodic_flush(self):
        self._pending_metrics = {}
        self._interval_secs = (
            self.config_manager.get_metric_config().get_interval_secs()
        )
        
        self.logger.info(f"Starting CloudWatch periodic flush with {self._interval_secs}s interval")
        
        self._flush_event = Event()
        self._flush_thread = Thread(
            target=self._flush_metrics_periodically, daemon=True
        )
        self._flush_thread.start()
        
        self.logger.debug("CloudWatch flush thread started")

    def close(self) -> None:
        """Stop the periodic-flush thread and flush any remaining metrics."""
        self._terminate_thread = True
        if self._flush_event is not None:
            self._flush_event.set()  # wake the pending wait()
        if self._flush_thread is not None:
            self._flush_thread.join(timeout=5)
            self._flush_thread = None
        # Best-effort final flush of anything still pending.
        try:
            self._flush_metrics()
        except Exception as e:
            self.logger.warning(f"Error during final CloudWatch flush: {e}")

    def _flush_metrics_periodically(self):
        while not self._flush_event.wait(self._interval_secs):
            # Never let an unexpected error kill the flush thread.
            try:
                self._flush_metrics()
            except Exception as e:
                self.logger.error(f"Unexpected error during CloudWatch flush: {e}")
            if self._terminate_thread:
                break

    def _flush_metrics(self):
        total_metrics = sum(len(metrics) for metrics in self._pending_metrics.values())
        if total_metrics == 0:
            self.logger.debug("No pending metrics to flush to CloudWatch")
            return

        self.logger.debug(f"Flushing {total_metrics} metrics across {len(self._pending_metrics)} namespaces to CloudWatch")

        for namespace, metrics in self._pending_metrics.items():
            if not metrics:
                continue
            # Isolate each namespace so one failing PutMetricData does not drop the
            # others, and chunk into <=1000-datum batches (CloudWatch's per-request
            # limit) so large batches are not rejected wholesale. On failure, keep
            # only the not-yet-sent datums so the next flush retries without
            # re-sending datums already accepted.
            remaining = []
            failed = False
            for start in range(0, len(metrics), self.MAX_DATUMS_PER_REQUEST):
                batch = metrics[start:start + self.MAX_DATUMS_PER_REQUEST]
                if failed:
                    remaining.extend(batch)
                    continue
                try:
                    self.logger.debug(f"Sending {len(batch)} metrics to CloudWatch namespace: {namespace}")
                    self._cloudwatch_client.put_metric_data(
                        Namespace=namespace, MetricData=batch
                    )
                except Exception as e:
                    self.logger.error(
                        f"Error sending metrics to CloudWatch namespace '{namespace}': {e}. "
                        f"Will retry on next flush."
                    )
                    failed = True
                    remaining.extend(batch)
            self._pending_metrics[namespace] = remaining

        self.logger.debug("CloudWatch flush cycle complete")

    def emit_metric(self, metric, measure_values):
        namespace = (
            metric.get_namespace()
            if metric.get_namespace is not None
            else self.config_manager.get_metric_config().get_namespace()
        )
        
        self.logger.debug(f"Queuing metric '{metric.get_name()}' for CloudWatch namespace: {namespace} with {len(measure_values)} measures")
        
        metric_data = self._prepare_metric_data(metric, measure_values)
        if namespace not in self._pending_metrics:
            self._pending_metrics[namespace] = []
        self._pending_metrics[namespace].extend(metric_data)
        
        total_pending = len(self._pending_metrics[namespace])
        self.logger.debug(f"Metric '{metric.get_name()}' queued for CloudWatch - {total_pending} metrics pending in namespace {namespace}")

    def emit_metric_now(self, metric, measure_values):
        namespace = (
            metric.get_namespace()
            if metric.get_namespace() is not None
            else self.config_manager.get_metric_config().get_namespace()
        )
        
        self.logger.debug(f"Emitting metric '{metric.get_name()}' immediately to CloudWatch namespace: {namespace} with {len(measure_values)} measures")
        
        metric_data = self._prepare_metric_data(metric, measure_values)
        self._cloudwatch_client.put_metric_data(
            Namespace=namespace, MetricData=metric_data
        )
        
        self.logger.debug(f"Metric '{metric.get_name()}' sent immediately to CloudWatch - {len(metric_data)} data points")

    def _prepare_metric_data(self, metric, measure_values):
        metric_data = []
        for measure_name, value in measure_values.items():
            data_point = {
                "MetricName": measure_name,
                "Dimensions": metric.dimensions_as_collection(False),
                "Timestamp": time.time(),
                "Value": value,
                "Unit": metric.get_measure(measure_name).get_unit(),
                "StorageResolution": metric.get_measure(
                    measure_name
                ).get_storage_resolution(),
            }
            metric_data.append(data_point)
            if self.metric_config.get_large_fleet_workaround():
                data_point = {
                    "MetricName": measure_name,
                    "Dimensions": metric.dimensions_as_collection(True),
                    "Timestamp": time.time(),
                    "Value": value,
                    "Unit": metric.get_measure(measure_name).get_unit(),
                    "StorageResolution": metric.get_measure(
                        measure_name
                    ).get_storage_resolution(),
                }
                metric_data.append(data_point)
        return metric_data

    def on_configuration_change(self, configuration) -> bool:
        old_interval = self._interval_secs
        new_interval = self.config_manager.get_metric_config().get_interval_secs()
        
        self.logger.info(f"CloudWatch configuration changed - interval: {old_interval}s -> {new_interval}s")
        
        # Stop current flush thread
        self.logger.debug("Stopping current CloudWatch flush thread")
        self._terminate_thread = True
        self._flush_thread.join()
        self._terminate_thread = False
        
        # Flush any pending metrics before reconfiguring
        self.logger.debug("Flushing pending metrics before reconfiguration")
        self._flush_metrics()
        
        # Start new flush thread with updated configuration
        self._start_periodic_flush()
        
        self.logger.info("CloudWatch target reconfiguration completed")
        return True

File: libs/java/src/main/java/com/mbreissi/edgecommons/metrics/targets/CloudWatch.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.mbreissi.edgecommons.metrics.targets;

import com.mbreissi.edgecommons.config.BufferConfiguration;
import com.mbreissi.edgecommons.config.ConfigManager;
import com.mbreissi.edgecommons.metrics.Metric;
import com.mbreissi.edgecommons.streaming.StreamHandle;
import com.mbreissi.edgecommons.streaming.StreamService;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import software.amazon.awssdk.services.cloudwatch.CloudWatchClient;
import software.amazon.awssdk.services.cloudwatch.model.CloudWatchException;
import software.amazon.awssdk.services.cloudwatch.model.Dimension;
import software.amazon.awssdk.services.cloudwatch.model.MetricDatum;
import software.amazon.awssdk.services.cloudwatch.model.PutMetricDataRequest;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

public final class CloudWatch extends MetricTarget
{

    /** The single callback stream name for the durable CloudWatch buffer. */
    static final String DURABLE_STREAM = "metrics-cw";

    private final CloudWatchClient cwClient;

    private final ConcurrentHashMap<String, ConcurrentLinkedQueue<PendingMetric>> pendingMetrics = new ConcurrentHashMap<>();

    private final ScheduledExecutorService scheduler =
            Executors.newSingleThreadScheduledExecutor(runnable -> {
                Thread thread = new Thread(runnable, "CloudWatch-emit-scheduler");
                thread.setDaemon(true);
                return thread;
            });
    private ScheduledFuture<?> emitTask;

    // ----- durable store-and-forward buffer (edgestreamlog Callback stream) -----
    private final boolean durable;
    private final CloudWatchDrain drain;     // null in memory mode
    private StreamService streamService;     // null in memory mode (or if durable init failed)
    private StreamHandle streamHandle;       // null in memory mode (or if durable init failed)

    public CloudWatch(ConfigManager configManager)
    {
        this(configManager, CloudWatchClient.builder().build());
    }

    /** Package-private constructor allowing a CloudWatch client to be injected for testing. */
    CloudWatch(ConfigManager configManager, CloudWatchClient cwClient)
    {
        super(configManager);
        this.cwClient = cwClient;

        BufferConfiguration buffer = metricConfig.getBufferConfig();
        boolean wantDurable = buffer != null && buffer.isDurable();
        boolean durableUp = false;
        CloudWatchDrain d = null;
        if (wantDurable)
        {
            if (!StreamService.nativeAvailable())
            {
                // Fail fast on an ABSENT native core. Durable is the default and is bundled by
                // design, so a missing core is a deployment error — silently degrading would lose
                // metrics across a cloud disconnect. Opt out explicitly with buffer.type=memory.
                throw new IllegalStateException(
                        "Durable CloudWatch buffer (the default) requires the edgestreamlog native "
                        + "core, which is not bundled/loadable for this platform. Bundle it (see "
                        + "docs/NATIVE_CORE_DELIVERY.md) or set "
                        + "metricEmission.targetConfig.cloudwatch.buffer.type=memory.");
            }
            // The drain owns the namespace grouping / stale-drop / chunk / outcome logic; the sink
            // forwards each PutMetricData chunk through the injected client. A core that is present
            // but whose buffer cannot open (e.g. a bad path) still degrades to in-memory below.
            d = new CloudWatchDrain(this::putChunk);
            durableUp = startDurable(buffer, d);
        }
        this.durable = durableUp;
        this.drain = durableUp ? d : null;

        if (!this.durable)
        {
            // In-memory path (explicit type=memory, or durable init failed → safe fallback).
            scheduleEmit();
        }
    }

    /**
     * Open the durable buffer: register the drain as the host sink, then open a single Callback
     * stream backed by a disk buffer at the (template-resolved) configured path. Returns false (and
     * the target falls back to in-memory batching) if the native streaming core is unavailable.
     */
    private boolean startDurable(BufferConfiguration buffer, CloudWatchDrain d)
    {
        try
        {
            // Resolve {ComponentName}/{ThingName} in the buffer path BEFORE opening — this also
            // closes the known Java streaming template-resolution gap for this path.
            String resolvedPath = configManager.resolveTemplate(buffer.getPath());
            String configJson = buildStreamConfig(resolvedPath, buffer);

            // Register the sink BEFORE open (the core binds it per stream at open time).
            StreamService.registerSink(batch -> d.drain(batch));
            this.streamService = StreamService.open(configJson);
            this.streamHandle = streamService.stream(DURABLE_STREAM);
            LOGGER.info("CloudWatch durable buffer open at {} (maxDiskBytes={}, onFull={}, fsync={})",
                    resolvedPath, buffer.getMaxDiskBytes(), buffer.getOnFull(), buffer.getFsync());
            return true;
        }
        catch (Throwable t)
        {
            LOGGER.warn("Durable CloudWatch buffer unavailable ({}); falling back to in-memory batching",
                    t.getMessage());
            closeDurable();
            return false;
        }
    }

    /** Build the single-stream {@code streaming} config JSON for the durable CloudWatch buffer. */
    private static String buildStreamConfig(String resolvedPath, BufferConfiguration buffer)
    {
        JsonObject sink = new JsonObject();
        sink.addProperty("type", "callback");
        sink.addProperty("id", DURABLE_STREAM);

        JsonObject buf = new JsonObject();
        buf.addProperty("type", "disk");
        buf.addProperty("path", resolvedPath);
        buf.addProperty("maxDiskBytes", buffer.getMaxDiskBytes());
        buf.addProperty("onFull", buffer.getOnFull());
        buf.addProperty("fsync", buffer.getFsync());

        JsonObject stream = new JsonObject();
        stream.addProperty("name", DURABLE_STREAM);
        stream.add("sink", sink);
        stream.add("buffer", buf);

        JsonArray streams = new JsonArray();
        streams.add(stream);
        JsonObject root = new JsonObject();
        root.add("streams", streams);
        return root.toString();
    }

    private void scheduleEmit()
    {
        // Reschedule on the same executor: cancel the current task and submit a new
        // one at the configured interval (executor reused; only shut down in close()).
        if (emitTask != null)
        {
            emitTask.cancel(false);
        }
        long periodMs = configManager.getMetricConfig().getIntervalSecs() * 1000L;
        // First flush after one interval (not at delay 0): nothing is buffered at
        // startup, so an immediate flush is pointless and only races with callers.
        emitTask = scheduler.scheduleAtFixedRate(this::runEmit, periodMs, periodMs, TimeUnit.MILLISECONDS);
    }

    @Override
    public void emitMetric(Metric metric, Map<String, Float> measureValues)
    {
        if (durable)
        {
            appendDurable(new PendingMetric(metric, measureValues));
            return;
        }
        // need to maintain a hashmap by namespace as we can only submit
        // from a single namespace at a time
        pendingMetrics.computeIfAbsent(metric.getNamespace(), k -> new ConcurrentLinkedQueue<>())
                     .add(new PendingMetric(metric, measureValues));
        LOGGER.info("Added {} metric to pending queue", metric.getName());
    }


    @Override
    public void emitMetricNow(Metric metric, Map<String, Float> measureValues)
    {
        if (durable)
        {
            // Durable mode: still goes through the buffer (always-buffer) — the engine drains it.
            appendDurable(new PendingMetric(metric, measureValues));
            return;
        }
        try
        {
            var data = new ArrayList<MetricDatum>();
            appendToPutMetricDataRequest(new PendingMetric(metric, measureValues), data);
            PutMetricDataRequest request = PutMetricDataRequest.builder()
                                                               .namespace(metric.getNamespace())
                                                               .metricData(data)
                                                               .build();
            cwClient.putMetricData(request);
        }
        catch (CloudWatchException e)
        {
            LOGGER.error("Error sending metric to CloudWatch. {}. Ignoring.", e.getMessage());
        }
        LOGGER.trace("Successfully sent {} metric to CloudWatch", metric.getName());
    }

    /** Serialize a pending metric's datums to {@code {namespace, datum}} records and append them. */
    private void appendDurable(PendingMetric pendingMetric)
    {
        var data = new ArrayList<MetricDatum>();
        appendToPutMetricDataRequest(pendingMetric, data);
        String namespace = pendingMetric.getMetric().getNamespace();
        for (MetricDatum datum : data)
        {
            try
            {
                byte[] payload = CloudWatchRecord.serialize(namespace, datum);
                // Partition key = namespace (so the drain can group by it).
                streamHandle.append(namespace, pendingMetric.getTimestamp().toEpochMilli(), payload);
            }
            catch (RuntimeException e)
            {
                LOGGER.error("Failed to append CloudWatch metric to durable buffer: {}", e.getMessage());
            }
        }
    }

    @Override
    public boolean onConfigurationChanged()
    {
        LOGGER.info("Configuration changed. Reconfiguring CloudWatch batch interval");
        if (!durable)
        {
            scheduleEmit();
        }
        return true;
    }

    @Override
    public void flush()
    {
        if (durable)
        {
            // Force the buffer durably to disk (the background engine handles the actual drain).
            try
            {
                if (streamHandle != null)
                {
                    streamHandle.flush();
                }
            }
            catch (RuntimeException e)
            {
                LOGGER.warn("Error flushing durable CloudWatch buffer: {}", e.getMessage());
            }
            return;
        }
        flushMetrics();
    }

    @Override
    public void close()
    {
        if (durable)
        {
            // Flush to disk + stop the engine; do NOT drain to cloud (backlog persists for restart).
            flush();
            closeDurable();
            StreamService.unregisterSink();
        }
        else
        {
            if (emitTask != null)
            {
                emitTask.cancel(false);
            }
            scheduler.shutdownNow();
        }
        try
        {
            cwClient.close();
        }
        catch (Exception e)
        {
            LOGGER.warn("Error closing CloudWatch client: {}", e.getMessage());
        }
    }

    private void closeDurable()
    {
        try
        {
            if (streamHandle != null)
            {
                streamHandle.close();
                streamHandle = null;
            }
        }
        catch (RuntimeException e)
        {
            LOGGER.warn("Error closing durable CloudWatch stream handle: {}", e.getMessage());
        }
        try
        {
            if (streamService != null)
            {
                streamService.close();
                streamService = null;
            }
        }
        catch (RuntimeException e)
        {
            LOGGER.warn("Error closing durable CloudWatch stream service: {}", e.getMessage());
        }
    }

    /** Datums dropped on drain for being outside CloudWatch's accept window (durable mode only). */
    public long getDroppedStale()
    {
        return drain == null ? 0L : drain.droppedStale();
    }

    /** The {@link CloudWatchSender} the durable drain calls — wraps the injected client. */
    private void putChunk(String namespace, List<MetricDatum> chunk)
    {
        sendBatch(namespace, chunk);
    }

    // CloudWatch PutMetricData accepts at most 1000 metric data items per request.
    private static final int MAX_DATUMS_PER_REQUEST = 1000;

    private void flushMetrics()
    {
        for (Map.Entry<String, ConcurrentLinkedQueue<PendingMetric>> entry : pendingMetrics.entrySet())
        {
            String namespace = entry.getKey();
            ConcurrentLinkedQueue<PendingMetric> pendingMetricQueue = entry.getValue();
            try
            {
                var data = new ArrayList<MetricDatum>();
                PendingMetric pendingMetric;
                while ((pendingMetric = pendingMetricQueue.poll()) != null)
                {
                    appendToPutMetricDataRequest(pendingMetric, data);
                }
                // CloudWatch PutMetricData accepts at most 1000 metric data items per call;
                // chunk the full datum list regardless of how it maps to individual metrics.
                for (int i = 0; i < data.size(); i += MAX_DATUMS_PER_REQUEST)
                {
                    sendBatch(namespace, data.subList(i, Math.min(i + MAX_DATUMS_PER_REQUEST, data.size())));
                }
            }
            catch (Exception e)
            {
                // Isolate failures per namespace so one failing namespace does not drop the others.
                LOGGER.error("Error sending pending metrics for namespace {} to CloudWatch. {}",
                        namespace, e.getMessage());
            }
        }
    }

    private void sendBatch(String namespace, Collection<MetricDatum> data)
    {
        PutMetricDataRequest request = PutMetricDataRequest.builder()
                                                           .namespace(namespace)
                                                           .metricData(data)
                                                           .build();
        cwClient.putMetricData(request);
        LOGGER.info("Successfully sent {} metric datums to CloudWatch namespace {}", data.size(), namespace);
    }

    private void appendToPutMetricDataRequest(PendingMetric pendingMetric, Collection<MetricDatum> data)
    {
        Map<String, Float> measureValues = pendingMetric.getMeasureValues();
        Collection<Dimension> dimensions = pendingMetric.getMetric().dimensionsAsCollection();
        Metric metric = pendingMetric.getMetric();
        Instant timestamp = pendingMetric.getTimestamp();

        addToDatum(data, measureValues, dimensions, metric, timestamp);

        if (metricConfig.getLargeFleetWorkaround())
        {
            dimensions = pendingMetric.getMetric().dimensionsAsCollection(true);
            addToDatum(data, measureValues, dimensions, metric, timestamp);
        }
    }

    private void addToDatum(Collection<MetricDatum> data, Map<String, Float> measureValues, Collection<Dimension> dimensions, Metric metric, Instant timestamp)
    {
        for (Map.Entry<String, Float> entry : measureValues.entrySet())
        {
            MetricDatum datum = MetricDatum.builder()
                                           .metricName(entry.getKey())
                                           .unit(metric.getMeasure(entry.getKey()).unit())
                                           .storageResolution(metric.getMeasure(entry.getKey()).storageResolution())
                                           .value(Double.valueOf(entry.getValue()))
                                           .timestamp(timestamp)
                                           .dimensions(dimensions)
                                           .build();
            data.add(datum);
        }
    }

    private static class PendingMetric
    {
        private final Metric metric;
        private final Map<String, Float> measureValues;

        private final Instant timestamp;

        public PendingMetric(Metric metric, Map<String, Float> measureValues)
        {
            this.metric = metric;
            this.measureValues = measureValues;
            this.timestamp = Instant.now();
        }

        public Metric getMetric()
        {
            return metric;
        }

        public Map<String, Float> getMeasureValues()
        {
            return measureValues;
        }

        public Instant getTimestamp()
        {
            return timestamp;
        }
    }

    /**
     * The periodic flush task. Guarded so an exception cannot propagate to the
     * scheduler and silently cancel future flushes.
     */
    private void runEmit()
    {
        try
        {
            flushMetrics();
        }
        catch (Exception e)
        {
            LOGGER.error("Unexpected error while emitting metrics to CloudWatch: {}", e.getMessage(), e);
        }
    }
}
